### PR TITLE
CA1028: EnumStorageShouldBeInt32 Analyzer

### PR DESCRIPTION
**Pull request by morrisjoe**
This contains only analyzer. Will send fixer in separate PR soon.

### CODE REVIEW

File: src/Microsoft.ApiDesignGuidelines.Analyzers/CSharp/Microsoft.ApiDesignGuidelines.CSharp.Analyzers.csproj [hunk 1]
```diff
@@ -25,6 +25,12 @@
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
   </PropertyGroup>
+  <PropertyGroup>
+    <DelaySign>false</DelaySign>
+  </PropertyGroup>
+  <PropertyGroup>
+    <SignAssembly>true</SignAssembly>
+  </PropertyGroup>
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/CSharp/Microsoft.ApiDesignGuidelines.CSharp.Analyzers.csproj]: Why this change?

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs [hunk 2]
```diff
@@ -25,25 +30,56 @@ public abstract class EnumStorageShouldBeInt32Analyzer : DiagnosticAnalyzer
...
+                var flagsAttribute = WellKnownTypes.FlagsAttribute(compilationContext.Compilation);
+                compilationContext.RegisterSymbolAction(symbolContext => AnalyzeSymbol(symbolContext, flagsAttribute), SymbolKind.NamedType);
+            });
+        }
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol flagsAttribute)
+        {
+            var symbol = (INamedTypeSymbol)context.Symbol;
+
+            // If not enum type exit
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: A nit-pick, but just one of my coding preferences: If the comment doesn't do anything except repeat what the code does, I would omit it. I add comments when it's necessary to describe _why_ the code does what it does, or when the implementation is unclear (but then I always ask if it could be clearer). In this case, I'd omit it.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs [hunk 3]
```diff
@@ -25,25 +30,56 @@ public abstract class EnumStorageShouldBeInt32Analyzer : DiagnosticAnalyzer
...
+                return;
+            }
+
+            // If enum is Int64 and has Flags attributes then exit
+            bool hasFlagsAttribute = symbol.GetAttributes().Any(a => a.AttributeClass.Equals(flagsAttribute));
+            if (underlyingType == SpecialType.System_Int64 && hasFlagsAttribute)
+            {
+                return;
+            }
+            else //Report diagnostic
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: No need for the `else` or the braces; just fall through (as you have done above after all the other tests).

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs [hunk 4]
```diff
@@ -4,20 +4,25 @@
...
-        private static readonly LocalizableString s_localizableMessageNotIntegral = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.EnumStorageShouldBeInt32MessageNotIntegral), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessageNotInt32 = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.EnumStorageShouldBeInt32Message), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
         private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.EnumStorageShouldBeInt32Description), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
         
         internal static DiagnosticDescriptor NotInt32Rule = new DiagnosticDescriptor(RuleId,
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: Since there's now only one rule in this analyzer, I'd rename `NotInt32Rule` to just `Rule`.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.Fixer.cs [hunk 5]
```diff
@@ -29,5 +29,13 @@ protected override CodeFixProvider GetCSharpCodeFixProvider()
...
             return new CSharpEnumStorageShouldBeInt32Fixer();
         }
+
+        #region CSharpUnitTests
+
+        #endregion
+
+        #region BasicUnitTests
+
+        #endregion
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.Fixer.cs]: My personal preference is to interleave the C# and VB tests so I don't have to look up and down the page to convince myself that every C# test has a matching VB test. (When I wrote that I was thinking of what I did when I was coding, but come to think of it, it makes it easier for me to review, as well, if each C# test is right next to its matching VB test.)

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 6]
```diff
@@ -12,12 +12,214 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+            public enum TestEnum4 : long //no violation - because resultant accessibility is private 
+            {
+                Value1 = 1,
+                Value2 = 2
+            }
+        }
+        public static void Main()
+        {
+        }
+    }
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: No need for the `Main` method. Yes, it's just test code, but I really feel that every extra line of code that doesn't serve a purpose should be removed. Lots of reasons for that which I can go on and on about :)

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 7]
```diff
@@ -12,12 +12,214 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+}
+ ");
+        }
+
+        [Fact]
+        public void CSharp_CA1028_Diagnostic1()
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Provide meaningful test method names. "Diagnostic1" doesn't tell you what the test is about when you're looking at a failure in Test Explorer or on the console output. For example: `CSharp_CA1028_DiagnosticForInt64WithNoFlags`.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 8]
```diff
@@ -12,12 +12,214 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+        {
+            VerifyCSharp(@"
+using System;
+namespace Test
+{
+    public enum TestEnum1 : long // violation - because underlying type is Int64 and has no Flags attribute
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: If the test name is meaningful you won't need these comments.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 9]
```diff
@@ -12,12 +12,214 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+        End Enum
+        Private Class innerClass
+            Public Enum TestEnum4 As Long 'no violation - because resultant accessibility Is private 
+                Value1 = 1
+                Value2 = 2
+            End Enum
+        End Class
+    End Class
+    Sub Main()
+    End Sub
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Remove `Main`.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 10]
```diff
@@ -12,12 +12,214 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+End Module
+ ");
+        }
+
+        [Fact]
+        public void Basic_CA1028_Diagnostic1()
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Again, meaningful test method names.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/VisualBasic/Microsoft.ApiDesignGuidelines.VisualBasic.Analyzers.vbproj [hunk 11]
```diff
@@ -24,6 +24,12 @@
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
   </PropertyGroup>
+  <PropertyGroup>
+    <DelaySign>false</DelaySign>
+  </PropertyGroup>
+  <PropertyGroup>
+    <SignAssembly>true</SignAssembly>
+  </PropertyGroup>
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/VisualBasic/Microsoft.ApiDesignGuidelines.VisualBasic.Analyzers.vbproj]: Why this change?

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 12]
```diff
@@ -12,12 +12,214 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+",
+    GetCA1028BasicResultAt(line: 4, column: 17, enumIdentifier: "TestEnum3", underlyingType: "UShort"));
+        }
+        #endregion
+
+        #region Helpers
+
+        private static DiagnosticResult GetCA1028CSharpResultAt(int line, int column, string enumIdentifier, string underlyingType)
+        {
+            // Format - Make the underlying type of {0} System.Int32 instead of {1}.
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: You don't have to do the formatting. There's an overload of `CreateDiagnostic` (or of the `Symbol` extension method I mentioned) that does it for you. See my proposed code above.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/CSharp/Microsoft.ApiDesignGuidelines.CSharp.Analyzers.csproj (hunk 1 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/CSharp/Microsoft.ApiDesignGuidelines.CSharp.Analyzers.csproj]: Good catch. Fixed.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs (hunk 2 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: Addressed.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs (hunk 3 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: Addressed.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs (hunk 4 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 6 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 7 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 9 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 10 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/VisualBasic/Microsoft.ApiDesignGuidelines.VisualBasic.Analyzers.vbproj (hunk 11 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/VisualBasic/Microsoft.ApiDesignGuidelines.VisualBasic.Analyzers.vbproj]: Good catch. Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 12 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: I am not clear on this comment. I am not seeing 'my proposed code above'.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 8 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Gave meaningful test names. But would still like to keep the comment

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 12 above)
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Huh, I don't see my comment that I was referring to. I must have failed to add it. I was saying that there's an easier way to create a diagnostic for a symbol. There's an extension method on `ISymbol` something like this (can't be bothered to look up the exact signature on my phone):

```
symbol.CreateDiagnostic(Rule, arg1, arg2...);
```

The symbol knows its location so you don't have to mention it. The rule knows its message so you don't have to extract it. The args get passed to `string.Format`, so you don't have to call `ToString` on any of them or call `string.Format` yourself.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs [hunk 13]
```diff
@@ -4,46 +4,80 @@
...
+
+            // If enum is Int64 and has Flags attributes then exit
+            bool hasFlagsAttribute = symbol.GetAttributes().Any(a => a.AttributeClass.Equals(flagsAttribute));
+            if (underlyingType == SpecialType.System_Int64 && hasFlagsAttribute)
+            {
+                return;
+            }
             
+            //Report diagnostic
+            var diagnostic = Diagnostic.Create(Rule, symbol.Locations[0], symbol.Name, symbol.EnumUnderlyingType.ToString());
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/EnumStorageShouldBeInt32.cs]: Here's where you can just say:
```
symbol.CreateDiagnostic(Rule, symbol.Name, symbol.EnumUnderlyingType);
```
Take a look at src\Analyzer.Utilities\Extensions.cs. You'll see lots of convenient extension methods named `CreateDiagnostic` or `CreateDiagnostics`. You can call `CreateDiagnostic` on a `SyntaxNode` as well as a symbol, because again, the node knows its own location.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs (hunk 12 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Addressed

Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Addressed

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs [hunk 14]
```diff
@@ -12,12 +12,191 @@ public class EnumStorageShouldBeInt32Tests : DiagnosticAnalyzerTestBase
...
+            VerifyBasic(@"
+Imports System
+Public Module Module1
+    Public Enum TestEnum2 As Byte 'violation - because underlying type is not Int32
+        Value1 = 1
+        Value2 = 2
+    End Enum
+    Sub Main()
+    End Sub
+End Module
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/EnumStorageShouldBeInt32Tests.cs]: Remove all the Sub Mains.

**Comment by lgolding** [comment]
:+1: except please remove all the occurrences of `Sub Main()`/`End Sub` from your VB test methods. They shouldn't be needed to make the tests pass. General rule: don't have code you don't need, for a variety of reasons.

I see you have a merge conflict. Ping me when you've resolved it and I'll take a last look.

**Comment by lgolding** [comment]
:+1:
